Refactoring - Camera (Capture and Video) UI Adaptation for Foldable Devices and Camera Initialization

yongsuk44 opened the pull request:
## To-do

- [x] : UI adjustment based on screen rotation and foldable state
- [ ] : Generating files to fit the capture and video screen size
- [ ] : Optimization and Refactoring

## Example screen

https://github.com/android/socialite/assets/38935359/48ee796e-7389-435e-ba15-955a4da726b0

romanofranz commented:
Thanks for doing this!
There are 2 issues when running it on the Pixel Fold (physical device): 
 - The camera viewfinder is sideways after rotation
 - The tabletop mode UI is not in sync with the fold orientation
![socialite-fold](https://github.com/android/socialite/assets/2760596/207a86fd-ba60-4223-8d84-44450578b1eb)

yongsuk44 commented:
@romanofranz 
Thanks, I'll check back. 😀

yongsuk44 commented:
@romanofranz 

I removed the code that locked the screen orientation when the orientation changes in the camera composable.

[Screen_recording.webm](https://github.com/android/socialite/assets/38935359/adf96784-02eb-4903-8d97-ee5de2d194ab)

romanofranz commented:
Thanks, this works better!
There are still few issues:
 - Performance issues on rotation (it takes ~1 second to render the preview after rotation). I changed the rotation animation to jumpcut to mitigate it, but we need to check the cause of the hiccup
 - If you rotate the device upside down (display rotation = 2), the preview is sideways. We may want to add a display listener to catch this corner case.

Overall LGTM on the camera side, adding @donovanfm to check the rest of the refactoring.
Good job!

yongsuk44 commented:
`android:screenOrientation` has been added to enable control of the Activity in four different orientations - [link](https://developer.android.com/guide/topics/manifest/activity-element#screen)

> - "fullSensor" : The device orientation sensor determines the orientation for any of the four orientations. This is similar to "sensor", except this allows for any of the four possible screen orientations regardless of what the device normally supports. For example, some devices don't normally use reverse portrait or reverse landscape, but this enables those orientations. Added in API level 9.
> - "user" : The user's current preferred orientation.
> - "fullUser"	: If the user has locked sensor-based rotation, this behaves the same as user, otherwise it behaves the same as fullSensor and allows any of the four possible screen orientations. Added in API level 18.

romanofranz reviewed (commented):
LGTM!

calren commented on app/src/main/java/com/google/android/samples/socialite/ui/camera/viewfinder/surface/Texture.kt:
can we remove this empty file

yongsuk44 commented on app/build.gradle.kts:
```diff
@@ -53,7 +53,11 @@ android {
         jvmTarget = "17"
-        freeCompilerArgs = listOf("-Xcontext-receivers")
+        freeCompilerArgs += listOf("-Xcontext-receivers")
+        freeCompilerArgs += listOf(
```
It wasn't related to the Camera UI.
Thank you for your feedback. 😀

Jaehwa-Noh commented on app/src/main/java/com/google/android/samples/socialite/ui/videoedit/VideoEditScreenViewModel.kt:
```diff
@@ -150,7 +149,7 @@ class VideoEditScreenViewModel @Inject constructor(
 
         val editedVideoFileName = "Socialite-edited-recording-" +
-            SimpleDateFormat(CameraViewModel.FILENAME_FORMAT, Locale.US)
+            SimpleDateFormat("yyyy-MM-dd-HH-mm-ss-SSS", Locale.US)
```
Is there any reason to remove FILENAME_FORMAT from CameraViewModel companion object and set format here?

yongsuk44 replied:
If you look at `CameraModule`, you can see that file names, paths, and types are created separately to be injected externally. In the future, When we refactor VideoEditScreen, we plan to change it to the same injection method and use separate constants to declare the format internally. 

Do you have any good suggestions? 🤔

Jaehwa-Noh replied:
You can do that simply making it as domain layer's usecase or just utils.

yongsuk44 replied:
Personally, I believe that it is appropriate to create UseCases only when handling the business logic of the program (such as chat, camera, video, etc.). On the other hand, providing only DateFormat as a Util seems like a good approach. 
I'll refactor and apply this after the current PR work is completed. Thank you for your suggestions :)

yongsuk44 commented on app/src/main/res/values/strings.xml:
```diff
@@ -66,4 +66,8 @@
     <string name="rw_title">Rewind</string>
     <string name="favorite_contact_widget_name">Favorite Contact</string>
 
+    <!--Camera Screen -->
+    <string name="photo">Photo</string>
+    <string name="video">Video</string>
```
Used by the `CameraControls` Composable. :)
